engine/execution/ingestion/engine.go

```diff
@@ -410,7 +410,42 @@ func (e *Engine) reloadBlock(
 		return fmt.Errorf("could not enqueue block %x on reloading: %w", blockID, err)
 	}
 
-	return e.addOrFetch(blockID, block.Header.Height, missingCollections)
+	height := block.Header.Height
```

Comment:
We can't reuse the `addOrFetch` function, because `addOrFetch` wants to acquire the lock of mempool, but reloadBlock has already acquired the same lock. Calling `addOrFetch` here would cause deadlock. 

So we had to duplicate the logic here in order to fetch collection and add directly the mempool data

Reply:
can you refactor addOrFetch instead of copying/pasting the logic?  Maybe have addOrFetch acquire the lock and call unsafeAddOrFetch, and use unsafeAddOrFetch here.

Reply:
Yeah, good point, but if we do that, then `addOrFetch` would hold the lock for too long. For instance, I will be holding the lock while calling the following methods:
- e.collection.ByID
- fetchCollection
- e.request.Force

I'd like to have `addOrFetch` holding the lock as minimum as possible, because holding for too long might cause other side effects.

Any other idea?

Reply:
I rather the code be correct than fast.  (There are a ton of thread-safe data races, aka byzantine failures, in the code base ...)

Which field(s)/method(s) does the mutex protect?  I'm not familiar with this code, but based on your list of methods, which I assume are data race free, the only thing you really need to guard is addCollectionToMempool. maybe restructure the code so that you only lock in addCollectionToMempool if that's the case.

Reply:
yeah, I refactored it a bit, and extracted the fetchAndHandleCollection function to be reused, which takes a function for handling collections, so that addAndFetch can provide a lock which acquires a lock, and reloadBlock can add the collection directly to the mempool as it's already holding the lock.